nemo/collections/nlp/data/machine_translation/preproc_mt_data.py

```diff
@@ -0,0 +1,406 @@
+from nemo.utils import logging
+
+
+class MTDataPreproc:
+    """Automatically preprocess machine translation.
+    """
+
+    def __init__(self, cfg: MTEncDecModelConfig, trainer: Trainer = None) -> None:
+        self._cfg = cfg
+        self.global_rank = 0
+        self.world_size = 1
+        if trainer is not None:
+            self.global_rank = (trainer.node_rank * trainer.num_gpus) + trainer.local_rank
+            self.world_size = trainer.num_nodes * trainer.num_gpus
+
+        if (
+            cfg.encoder_tokenizer.get('tokenizer_name') != 'yttm'
+            or cfg.decoder_tokenizer.get('tokenizer_name') != 'yttm'
+        ):
+            raise NotImplementedError(f"Currently we only support yttm tokenizer.")
+
+        # Train tokenizer models if they don't exist
+        if (
+            cfg.encoder_tokenizer.get('tokenizer_model') is None
+            or cfg.decoder_tokenizer.get('tokenizer_model') is None
+        ):
+            if cfg.get('preproc_out_dir') is None:
+                raise ValueError('Tokenizer model training required but cfg.preproc_out_dir is None.')
+            # train tokenizer model on training data
+            self.encoder_tokenizer_model, self.decoder_tokenizer_model = self.train_tokenizers(
+                out_dir=cfg.get('preproc_out_dir'),
+                src_fname=cfg.train_ds.get('src_file_name'),
+                tgt_fname=cfg.train_ds.get('tgt_file_name'),
+                shared_tokenizer=cfg.get('shared_tokenizer'),
+                encoder_tokenizer_vocab_size=cfg.encoder_tokenizer.get('vocab_size'),
+                decoder_tokenizer_vocab_size=cfg.decoder_tokenizer.get('vocab_size'),
+                encoder_tokenizer_name=cfg.encoder_tokenizer.get('tokenizer_name'),
+                decoder_tokenizer_name=cfg.decoder_tokenizer.get('tokenizer_name'),
+            )
+            # update config
+            self._cfg.encoder_tokenizer.tokenizer_model = self.encoder_tokenizer_model
+            self._cfg.decoder_tokenizer.tokenizer_model = self.decoder_tokenizer_model
+        else:
+            self.encoder_tokenizer_model = cfg.encoder_tokenizer.get('tokenizer_model')
+            self.decoder_tokenizer_model = cfg.decoder_tokenizer.get('tokenizer_model')
+
+        self.encoder_tokenizer, self.decoder_tokenizer = self.get_enc_dec_tokenizers(
+            encoder_tokenizer_name=cfg.encoder_tokenizer.get('tokenizer_name'),
+            encoder_tokenizer_model=self.encoder_tokenizer_model,
+            encoder_bpe_dropout=cfg.encoder_tokenizer.get('bpe_dropout', 0.0),
+            decoder_tokenizer_name=cfg.decoder_tokenizer.get('tokenizer_name'),
+            decoder_tokenizer_model=self.decoder_tokenizer_model,
+            decoder_bpe_dropout=cfg.decoder_tokenizer.get('bpe_dropout', 0.0),
+        )
+
+        # If using tarred dataset for training, automatically create it if needed
+        if hasattr(cfg, 'train_ds'):
+            if cfg.train_ds.get('use_tarred_dataset'):
```

Comment:
Maybe also check for the case `cfg.train_ds is None` and print error message in that case instead of crashing

Reply:
It shouldn't crash if train_ds is none, rather just do nothing.

I did address the issue if tokenizer model is missing but there are no src and tgt files to train on.